### PR TITLE
Lido app/marcus/unique

### PR DESCRIPTION
**Pull request by marhong2**
@tyleradams

### CODE REVIEW

File: formulas/functions/reference.js [hunk 1]
```diff
@@ -288,6 +288,40 @@ const ReferenceFunctions = {
...
+   *          Rows are returned in the order in which they first appear in the source range.
+   **/
+  UNIQUE: (range) => {
+    try {
+      range = H.accept(range, Types.ARRAY, null, false);
+    } catch (e) {
+      range = H.accept(range);
+      return range;
+    }
+    if (typeof range[0] !== "object") {
```
Comment (tyleradams) [inline formulas/functions/reference.js:303]: Add a comment saying why we need this check, I have no idea

File: formulas/functions/reference.js [hunk 2]
```diff
@@ -288,6 +288,40 @@ const ReferenceFunctions = {
...
+    }
+    let seen = new Set();
+    let rv = [];
+    for (let index = 0; index < range.length; index++) {
+      let currValue = range[index];
+      let returnValue = currValue;
+      currValue = H.accept(currValue, Types.ARRAY);
+      //Since JS compares based upon pointers, converting our Arrays to strings allows for element wise comparisons.
+      try {
+        currValue = currValue.join();
```
Comment (tyleradams) [inline formulas/functions/reference.js:314]: Why not check that currValue has join defined?

File: test/formulas/reference/testcase.js [hunk 3]
```diff
@@ -104,6 +104,16 @@ module.exports = {
         'SUM(TRANSPOSE({1,2,3;4,5,6}))': 21,
     },
 
+    UNIQUE: {
+        'SUM(UNIQUE(C2:C10))': 176,
```
Comment (tyleradams) [inline test/formulas/reference/testcase.js:108]: Could we do something like `COUNT(UNIQUE(C2:C10))`?

**Review by tyleradams — CHANGES_REQUESTED** [review]

File: formulas/functions/reference.js (hunk 1 above)
Comment (tyleradams) [reply formulas/functions/reference.js:305]: Is checking object sufficient for subarray? Object feels too generic.

**Review by tyleradams — APPROVED** [review]

File: test/formulas/reference/testcase.js [hunk 4]
```diff
@@ -104,6 +104,26 @@ module.exports = {
...
+        'SUM(UNIQUE(D1:E12))': 20,
+        'SUM(UNIQUE(B2:E10))': 326.77,
+        'SUM(UNIQUE(D1))' : 4,
+        'SUM(UNIQUE(D10:E12))': 11,
+        'SUM(UNIQUE(C2:E2))': 51,
+        'COUNT(UNIQUE(D2:D10))': 1,
+        'COUNT(UNIQUE(B2:C10))': 18,
+        'COUNT(UNIQUE(B2:E10))': 36,
+        'COUNT(UNIQUE(D10:E12))': 4,
+        'COUNT({38,15,25,10,16,20,12})': 7,
```
Comment (tyleradams) [inline test/formulas/reference/testcase.js:119]: Add a case calling unique with duplicates.

File: formulas/functions/reference.js [hunk 5]
```diff
@@ -372,6 +406,200 @@ const ReferenceFunctions = {
...
+   *                          1 - Exact match. If none found, return the next larger item.
+   *                          2 - A wildcard match where *, ?, and ~ have special meaning.
+   * @param search_mode: OPTIONAL:
+   *                          1 - Perform a search starting at the first item. This is the default.
+   *                         -1 - Perform a reverse search starting at the last item.
+   *                          2 - Perform a binary search that relies on lookup_array being sorted in ascending order. If not sorted, invalid results will be returned.
+   *                         -2 - Perform a binary search that relies on lookup_array being sorted in descending order. If not sorted, invalid results will be returned.
+   * Microsoft Link: https://support.microsoft.com/en-us/office/xlookup-function-b7fd680e-6d10-43e6-84f9-88eae8bf5929
+   ***/
+  XLOOKUP: (
```
Comment (tyleradams) [inline formulas/functions/reference.js:426]: This can't be in this PR

**Review by tyleradams — CHANGES_REQUESTED** [review]

File: formulas/functions/reference.js (hunk 5 above)
Comment (marhong2) [reply formulas/functions/reference.js:426]: A side effect of my squashes, I have been trying to fix it

**Pull request closed without merging**